theseus/src/api/mod_extraction.rs

```diff
@@ -0,0 +1,97 @@
+use std::error::Error;
+use std::fs::{self, File};
+use std::path::{Path, PathBuf};
+use std::io::BufReader;
+use std::time::SystemTime;
+use serde_json::Value;
+use zip::ZipArchive;
+use zip::read::ZipFile;
+
+use crate::model::mod_type::{JARLoadedMod, ModConfig};
+use crate::util::fetch::sha1_async;
+
+/// Extract meta-data from JAR file.
+#[tracing::instrument]
+pub async fn extract_info_from_jar(file_path: String) -> crate::Result<JARLoadedMod> {
+    let path = Path::new(&file_path);
+
+    //let path_str = path.as_os_str().to_str();
+    let jar_buf = std::fs::read(path)?;
+    let jar_bytes = bytes::Bytes::from(jar_buf);
+    let jar_hash = sha1_async(jar_bytes).await;
+    let timestamp_added = get_file_datetime(path);
+
+    let mod_config = read_config_from_jar(path).expect("Failed to extract JSON from JAR");
+    
+    Ok(JARLoadedMod {
+        file_hash: jar_hash,
+        absolute_path: file_path,
+        timestamp_added: timestamp_added,
+        mod_config: mod_config,
+    })
+}
+
+// Get timestamp of the file being modified.
+fn get_file_datetime(file_path: &Path) -> Option<SystemTime> {
+    fs::metadata(file_path).ok().map(|metadata| metadata.modified())?.ok()
+}
+
+fn read_config_from_jar(jar_file: &Path) -> Result<ModConfig, String> {
```

Comment:
Using a `String` for an error message is bad/not what we want to do here. If an error isn't in the main API struct definition, you should add it and use thiserror to create error messages accordingly

Reply:
Fixed, thanks